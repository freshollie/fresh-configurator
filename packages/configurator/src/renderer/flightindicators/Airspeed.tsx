import React from "react";

import { SpeedMechanics, FiNeedle } from "./assets";
import withBase from "./hoc/withBase";

const AIRSPEED_LIMIT_H = 160;
const AIRSPEED_LIMIT_L = 0;

const boundedSpeed = (speed: number): number =>
  Math.max(AIRSPEED_LIMIT_L, Math.min(AIRSPEED_LIMIT_H, speed));

export default withBase<{
  speed: number;
}>(({ speed }) => (
  <>
    <SpeedMechanics />
    <FiNeedle
      data-testid="indicator-needle"
      style={{ transform: `rotate(${90 + boundedSpeed(speed) * 2}deg)` }}
    />
  </>
));
