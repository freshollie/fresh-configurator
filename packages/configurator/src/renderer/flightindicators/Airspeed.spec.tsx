import React from "react";
import { render } from "@testing-library/react";
import Airspeed from "./Airspeed";

describe("Airspeed", () => {
  it("should render correctly", () => {
    const { asFragment } = render(<Airspeed speed={100} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("should rotate the airspeed needle to match the speed", () => {
    const { getByTestId, rerender } = render(<Airspeed speed={50} />);

    expect(getByTestId("indicator-needle")).toHaveStyle(
      "transform: rotate(190deg)"
    );

    rerender(<Airspeed speed={80} />);
    expect(getByTestId("indicator-needle")).toHaveStyle(
      "transform: rotate(250deg)"
    );
  });

  it("should limit airspeed to 160", () => {
    const { getByTestId, rerender } = render(<Airspeed speed={160} />);

    expect(getByTestId("indicator-needle")).toHaveStyle(
      "transform: rotate(410deg)"
    );

    rerender(<Airspeed speed={300} />);
    expect(getByTestId("indicator-needle")).toHaveStyle(
      "transform: rotate(410deg)"
    );
  });

  it("should ignore speed less than 0", () => {
    const { getByTestId, rerender } = render(<Airspeed speed={0} />);

    expect(getByTestId("indicator-needle")).toHaveStyle(
      "transform: rotate(90deg)"
    );

    rerender(<Airspeed speed={-50} />);
    expect(getByTestId("indicator-needle")).toHaveStyle(
      "transform: rotate(90deg)"
    );
  });
});
