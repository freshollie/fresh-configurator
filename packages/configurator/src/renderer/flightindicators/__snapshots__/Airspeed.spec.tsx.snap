// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Airspeed should render correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-2 {
  width: 250px;
  height: 250px;
  position: relative;
  display: inline-block;
  overflow: hidden;
}

.emotion-2 div,
.emotion-2 svg {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

<div
    class="emotion-0 emotion-1 emotion-2 bb-Box"
  >
    <svg
      data-asset="SpeedMechanics"
    />
    <svg
      data-asset="FiNeedle"
      data-testid="indicator-needle"
      style="transform: rotate(290deg);"
    />
    <svg
      data-asset="FiCircle"
    />
  </div>
</DocumentFragment>
`;
